src/lara-app/components/runtime.tsx

```diff
@@ -79,34 +84,50 @@ export const RuntimeComponent = ({experiment, runKey, firebaseJWT, setError, def
 
   // get experiment data when loaded
   useEffect(() => {
-    firebase
-      .firestore()
-      .collection(`runs/${runKey}/experiments`)
-      .orderBy("createdAt", "desc")
-      .limit(1)
-      .onSnapshot(snapshot => {
-        setQueriedFirestore(true);
-        experimentRef.current = snapshot.docs[0]?.data?.().experiment;
-        const newData = experimentRef.current?.data as IExperimentData | undefined;
-        setExperimentData(newData);
-
-        // if there is no data force an upload
-        setDisplayQrAndMaybeRegenerateQR((newData === undefined) && !reportMode);
-      }, (err) => {
-        setError(err);
-      });
+    if (runKey) {
+      firebase
+        .firestore()
+        .collection(`runs/${runKey}/experiments`)
+        .orderBy("createdAt", "desc")
+        .limit(1)
+        .onSnapshot(snapshot => {
+          setQueriedFirestore(true);
+          experimentRef.current = snapshot.docs[0]?.data?.().experiment;
+          const newData = experimentRef.current?.data as IExperimentData | undefined;
+          setExperimentData(newData);
+
+          // if there is no data force an upload
+          setDisplayQrAndMaybeRegenerateQR((newData === undefined) && !reportOrPreviewMode);
+        }, (err) => {
+          setError(err);
+        });
+    }
   }, []);
 
   // re-generate QR code if showing
   useEffect(() => {
     const interval = setInterval(() => {
-      if (displayingCode.current && !reportMode) {
-        generateQRCode();
+      if (displayingCode.current && !reportOrPreviewMode && runKey && firebaseJWT) {
+        generateQRCode({runKey, firebaseJWT});
```

Comment:
Out of curiosity - you've added arguments to `generateQRCode` because there were some issues or just to prevent them? I would say there's not much difference and runKey and firebaseJWT would be frozen to the initial render values anyway, but I might be lost in closures. :wink:

Reply:
I did that to remove a conditional inside the function since both runKey and firebaseJWt can be undefined in the outer context.